src/kbc_modules/uri.rs

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2023 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+//! Resource Id is the url of KBS URL.
+
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+const RESOURCE_ID_ERROR_INFO: &str =
+    "invalid kbs uri, should be kbs://<addr-of-kbs>/<repo>/<type>/<tag>";
+
+const SCHEME: &str = "kbs";
+
+/// Resource Id is proposed in <https://github.com/confidential-containers/documentation/issues/85>
+#[derive(Clone, Debug, PartialEq)]
+pub struct ResourceId {
+    pub kbs_addr: String,
+    pub repository: String,
+    pub r#type: String,
+    pub tag: String,
+}
+
+impl TryFrom<&str> for ResourceId {
+    type Error = &'static str;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
```

Comment:
Perhaps the implementation of this function can be simplified into the following two lines?
```rust
let url = url::Url::try_from(value)?;
Self::try_from(url)
```